### PR TITLE
linkcheck: Do not check the availability of the same URL repeatedly

### PR DESCRIPTION
**Pull request by tk0miya**
### Feature or Bugfix
- Bugfix

### Purpose
- So far, linkcheck scans all of references and images from documents, and
checks them all simply.
- This collects the URL via post-transforms, and removes duplicated URLs
before checking availability.
- refs: #4303
  * Note: This does not change the behavior for the different anchors to the same page.

### CODE REVIEW

**Comment by francoisfreitag** [comment]
Great idea :+1: 

> So far, linkcheck scans all of references and images from documents, and checks them all simply.

I believe the `linkcheck` builder remembers previously checked URLs and verifies if an URL was already checked before to re-check it. However, the current feature is sensitive to race conditions, an URL could be checked multiple times by several threads, until one of them finishes and caches the result.

The suggested approach fixes that race condition by removing the duplicate links before threads start checking them :tada:. I believe we can remove the old “cache”: https://github.com/tk0miya/sphinx/pull/6.

---
### Follow-up ideas

There are preconditions to check a link: https://github.com/sphinx-doc/sphinx/blob/3ed7590ed411bd93b26098faab4f23619cdb2267/sphinx/builders/linkcheck.py#L238-L244

Instead of queuing these links, perhaps they should be skipped by the hyperlink collector?

Similarly, the `self.to_ignore` built from the `linkcheck_ignore` config option, though these need to be logged.

Moving these checks outside of the threads would help keeping the `check_thread` function (and threads) more focused on checking URIs, which would improve the current architecture IMO.

I can make these changes in separate follow-up PRs if you agree?

File: tests/roots/test-linkcheck-localserver-two-links/index.rst [hunk 1]
```diff
@@ -1,6 +1,6 @@
-.. image:: http://localhost:7777/
+.. image:: http://localhost:7777/image.png
    :target: http://localhost:7777/
 
 `weblate.org`_
 
-.. _weblate.org: http://localhost:7777/
+.. _weblate.org: http://localhost:7777/weblate
```
Comment (francoisfreitag) [inline tests/roots/test-linkcheck-localserver-two-links/index.rst]: With all links being different, all items in the priority queue are now comparable (the uri is the second element of Hyperlink, right after the `next_check` and this change guarantees uris are unique in the queue). We can revert a1b8b1febb6ce0b7423bb78df30b0fc424c12325 :grin:

**Comment by tk0miya** [comment]
>There are preconditions to check a link:

Good to me. But the skipped items should be output to the log. So we need to consider how to do that.

File: tests/roots/test-linkcheck-localserver-two-links/index.rst (hunk 1 above)
Comment (tk0miya) [reply tests/roots/test-linkcheck-localserver-two-links/index.rst]: I understand what this test does. Okay, let's remove (or deprecate) them all.

File: CHANGES [hunk 2]
```diff
@@ -57,6 +58,7 @@ Bugs fixed
   + or ^) are used as keystrokes
 * #8629: html: A type warning for html_use_opensearch is shown twice
 * #8665: html theme: Could not override globaltoc_maxdepth in theme.conf
+* #4304: linkcheck: Do not check the availability of the same URL repeatedly
```
Comment (francoisfreitag) [inline CHANGES]: This entry is misleading. It implies the availability was being checked repeatedly, when in most cases it wasn’t.

Perhaps:

Fix race condition that could lead to checking the availability of the same URL twice

**Review by francoisfreitag — APPROVED** [review]
Otherwise LGTM.

File: CHANGES (hunk 2 above)
Comment (tk0miya) [reply CHANGES]: Indeed!

**Comment by tk0miya** [comment]
@francoisfreitag Thank you for reviewing!

**Pull request merged**